### PR TITLE
Avoid error when bind mount source already exists

### PR DESCRIPTION
**Pull request by agners**

### CODE REVIEW

File: src/core/mount.c [hunk 1]
```diff
@@ -1010,7 +1010,7 @@ static void mount_enter_mounting(Mount *m) {
         p = get_mount_parameters_fragment(m);
         if (p && mount_is_bind(p)) {
                 r = mkdir_p_label(p->what, m->directory_mode);
-                if (r < 0)
+                if (r < 0 && r != -EEXIST)
```
Comment (bluca) [inline src/core/mount.c:1013]: Mmmh mkdir_p & family already return 0 if there's an EEXIST along the way, as far as I can see. What problem are you seeing precisely, and how to reproduce it?

**Comment by poettering** [comment]
yes, mkdir_p() is a NOP on EEXIST, so this change should be redundant. wha are you trying to fix?

**Comment by poettering** [comment]
Let's close this, the patch is redundant, as mentioned. Please report an issue with the issue you ran into first, and we can have a closer look.

**Comment by hundeboll** [comment]
@poettering We are getting this error when bind mounting a file, so we should probably add a type check for `p->what` before calling `mkdir_p_label()` or check the returned error like here.

**Comment by hundeboll** [comment]
Something like this (not tested):
```patch
diff --git a/src/core/mount.c b/src/core/mount.c
index 5479cf7bf2..e867c867c1 100644
--- a/src/core/mount.c
+++ b/src/core/mount.c
@@ -26,6 +26,7 @@
 #include "process-util.h"
 #include "serialize.h"
 #include "special.h"
+#include "stat-util.h"
 #include "string-table.h"
 #include "string-util.h"
 #include "strv.h"
@@ -1008,12 +1009,19 @@ static void mount_enter_mounting(Mount *m) {
 
         /* Create the source directory for bind-mounts if needed */
         p = get_mount_parameters_fragment(m);
-        if (p && mount_is_bind(p)) {
+        if (p && mount_is_bind(p) && access(m->where, F_OK) >= 0 && is_dir(m->where, true) > 0) {
                 r = mkdir_p_label(p->what, m->directory_mode);
                 if (r < 0)
                         log_unit_error_errno(UNIT(m), r, "Failed to make bind mount source '%s': %m", p->what);
         }
 
+        /* Create the target file for bind-mounts if needed */
+        if (p && mount_is_bind(p) && access(p->what, F_OK) >= 0 && access(m->where) < 0 && is_dir(p->what, true) == 0) {
+                r = touch(m->where);
+                if (r < 0)
+                        log_unit_error_errno(UNIT(m), r, "Failed to make bind mount target '%s': %m", m->where);
+        }
+
         if (p) {
                 _cleanup_free_ char *opts = NULL;
 
```

**Comment by hundeboll** [comment]
Grabbed the opportunity to also address https://github.com/systemd/systemd/issues/17184

**Comment by poettering** [comment]
> @poettering We are getting this error when bind mounting a file, so we should probably add a type check for `p->what` before calling `mkdir_p_label()` or check the returned error like here.

"this error"? Which error precisely? please paste the precise error message you saw?

**Comment by hundeboll** [comment]
The error log made conditional in this PR:
`Jan 07 11:59:10 stim-v1 systemd[1]: etc-fw_env.config.mount: Failed to make bind mount source '/etc/fw_env.stim': File exists`

**Comment by poettering** [comment]
ah, i get it now. so you have the inode to overmount already in place as a regular file, and then you want to overmount it, but we complain because mkdir_p() will return EEXIST if called on an existing inode that is not a dir. i see now.

I think the original patch actually does make sense then. sorry for closing this prematurely.

Can you add a comment to the sources though that explains that EEXIST can happen if the path exists and is not a dir? I didn't realize that part, and hence didn't grok things. And i have the suspicion that the next one looking at this will be confused by that too, hence a comment would clear things up.

**Comment by poettering** [comment]
Would be happy to review a fix for #17184, but please submit as follow-up PR. Also, please make sure we minimize conditionalization and only check ones whether source inode is dir or not.

**Comment by poettering** [comment]
#18200 has since been merged. I think fixing #17184 would be excellent. Would love to review/merge a patch for that. But let's do that in a new PR

**Pull request closed without merging**